lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -688,6 +692,51 @@ impl Instance {
         self.c_fatal_handler = Some(handler);
     }
 
+    pub fn begin_hostcall(&mut self) {
+        // Lock the current execution domain, so we can update to `Hostcall`.
+        let mut current_domain = self.kill_state.execution_domain();
+        match *current_domain {
+            Domain::Guest => {
+                // Guest is the expected domain until this point. Switch to the Hostcall
+                // domain so we know to not interrupt this instance.
+                *current_domain = Domain::Hostcall;
+            }
+            Domain::Hostcall => {
+                panic!(
+                    "Invalid state: Instance marked as in a hostcall while entering a hostcall."
+                );
+            }
+            Domain::Terminated => {
+                panic!("Invalid state: Instance marked as terminated while in guest code. This should be an error.");
+            }
+        }
+    }
+
+    pub fn end_hostcall(&mut self) {
+        let mut current_domain = self.kill_state.execution_domain();
+        match *current_domain {
+            Domain::Guest => {
+                panic!("Invalid state: Instance marked as in guest code while exiting a hostcall.");
+            }
+            Domain::Hostcall => {
+                *current_domain = Domain::Guest;
+            }
+            Domain::Terminated => {
+                // The instance was stopped in the hostcall we were executing.
+                std::mem::drop(current_domain);
+                unsafe {
+                    self.terminate(TerminationDetails::Remote);
```

Comment:
We need to remember to change this to a panic once the unwinding support is merged, as this currently will otherwise drop frames.